ZenPacks/zenoss/OpenStackInfrastructure/apiclients/session.py

```diff
@@ -0,0 +1,516 @@
+        log.debug('curl "{url}" {headerargs}'.format(
+            url=url,
+            headerargs=headerargs
+        ))
+
+        try:
+            response = yield semaphore.run(
+                add_timeout,
+                agent.request(
+                    'GET',
+                    url,
+                    headers=headers
+                ),
+                READ_TIMEOUT
+            )
+        except TimeoutError:
+            raise TimeoutError("GET %s" % url)
+        body = yield readBody(response)
+
+        # If the request resulted in an error, raise an exception
+        self.handle_error_response(response, body)
+
+        returnValue((body, response.headers))
+
+    @inlineCallbacks
+    def authenticated_GET_request(self, url, headers=None, params=None):
+        if headers is None:
+            headers = self.default_headers()
+
+        if self.token_id is None:
+            yield self.authenticate()
+        headers.setRawHeaders('X-Auth-Token', [self.token_id])
+
+        try:
+            result = yield self.GET_request(url, headers=headers, params=params)
+            returnValue(result)
+        except UnauthorizedError:
+            log.info("Received unauthorized response to http request, re-authenticating.  (%s)", url)
+            yield self.authenticate()
+            headers.setRawHeaders('X-Auth-Token', [self.token_id])
+
+        result = yield self.GET_request(url, headers=headers, params=params)
+        returnValue(result)
+
+    @inlineCallbacks
+    def POST_request(self, url, headers=None, body=None):
+        scheme = urlparse(url).scheme
+        agent = self.agent(scheme)
+        if headers is None:
+            headers = self.default_headers()
+
+        if body is None:
+            body = ""
+
+        semaphore = getDeferredSemaphore(self.auth_url, MAX_PARALLEL)
+
+        log.debug("POST %s", url)
+
+        try:
+            response = yield semaphore.run(
+                add_timeout,
+                agent.request(
+                    'POST',
+                    url,
+                    headers=headers,
+                    bodyProducer=StringProducer(body)
+                ),
+                READ_TIMEOUT
+            )
+        except TimeoutError:
+            raise TimeoutError("POST %s" % url)
+
+        body = yield readBody(response)
+
+        # If the request resulted in an error, raise an exception
+        self.handle_error_response(response, body)
+
+        returnValue((body, response.headers))
+
+    @inlineCallbacks
+    def authenticated_POST_request(self, url, headers=None, body=None):
+        if self.token_id is None:
+            yield self.authenticate()
+
+        try:
+            result = yield self.POST_request(url, headers=headers, body=body)
+            returnValue(result)
+        except UnauthorizedError:
+            log.info("Received unauthorized response to http request, re-authenticating.  (%s)", url)
+            yield self.authenticate()
```

Comment:
The idea is that the only real situation where we're going to get an unauthorized error after having successfully authenticated is that our token has expired- so it should only take one attempt to get a new one and be good to go.  More than that, and something's more broken, and recovery isn't likely, so i'd rather just let it fail.